src/core/ext/client_channel/retry_throttle.c

```diff
@@ -0,0 +1,242 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "src/core/ext/client_channel/retry_throttle.h"
+
+#include <limits.h>
+#include <string.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/atm.h>
+#include <grpc/support/avl.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/sync.h>
+
+//
+// server_retry_throttle_data
+//
+
+struct grpc_server_retry_throttle_data {
+  gpr_refcount refs;
+  int max_milli_tokens;
+  int milli_token_ratio;
+  gpr_atm milli_tokens;
+  // A pointer to the replacement for this grpc_server_retry_throttle_data
+  // entry.  If non-NULL, then this entry is stale and must not be used.
+  // We hold a reference to the replacement.
+  gpr_atm replacement;
+};
+
+static void get_replacement_throttle_data_if_needed(
+    grpc_server_retry_throttle_data** throttle_data) {
+  while (true) {
+    grpc_server_retry_throttle_data* new_throttle_data =
+        (grpc_server_retry_throttle_data*)gpr_atm_acq_load(
+            &(*throttle_data)->replacement);
+    if (new_throttle_data == NULL) return;
+    // Reset *throttle_data to its replacement, updating refcounts as
+    // appropriate.
+    // Note: It's safe to do this here, because the caller ensures that
+    // this will only be called with a given value of throttle_data from
+    // one thread at a time.
+    grpc_server_retry_throttle_data_ref(new_throttle_data);
+    grpc_server_retry_throttle_data* old_throttle_data = *throttle_data;
+    *throttle_data = new_throttle_data;
+    grpc_server_retry_throttle_data_unref(old_throttle_data);
+  }
+}
+
+bool grpc_server_retry_throttle_data_record_failure(
+    grpc_server_retry_throttle_data** throttle_data) {
+  // First, check if we are stale and need to be replaced.
+  get_replacement_throttle_data_if_needed(throttle_data);
+  // We decrement milli_tokens by 1000 (1 token) for each failure.
+  const int delta = -1000;
+  const int old_value = (int)gpr_atm_full_fetch_add(
+      &(*throttle_data)->milli_tokens, (gpr_atm)delta);
+  // If the above change takes us below 0, then re-add the excess.  Note
+  // that between these two atomic operations, the value will be
+  // artificially low by as much as 1000, but this window should be
+  // brief.
+  int new_value = old_value - 1000;
+  if (new_value < 0) {
+    const int excess_value = new_value - (old_value < 0 ? old_value : 0);
+    gpr_atm_full_fetch_add(&(*throttle_data)->milli_tokens,
```

Comment:
Would it make sense to do an atomic store of zero here? Otherwise aren't race conditions introduced in which the milli_tokens data is mutated since the original gpr_atm_full_fetch_add call?